### PR TITLE
ExpressionProjection add parameters field to support expression aggregation

### LINKED ISSUE #17402: Doesn't support IFNULL ?
### Which version of ShardingSphere did you use?
5.1.1

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
ShardingSphere-Proxy

# max():
SQL:
```
MySQL [hireez_sharding_db]> select max(email_campaign_account_id)  from EMAIL_CAMPAIGN;
+--------------------------------+
| max(email_campaign_account_id) |
+--------------------------------+
|                           NULL |
+--------------------------------+
1 row in set (0.010 sec)
```
log:
```
[INFO ] 2022-05-07 02:11:25.914 [ShardingSphere-Command-16] ShardingSphere-SQL - Logic SQL: select max(email_campaign_account_id)  from EMAIL_CAMPAIGN
[INFO ] 2022-05-07 02:11:25.914 [ShardingSphere-Command-16] ShardingSphere-SQL - SQLStatement: MySQLSelectStatement(table=Optional.empty, limit=Optional.empty, lock=Optional.empty, window=Optional.empty)
[INFO ] 2022-05-07 02:11:25.914 [ShardingSphere-Command-16] ShardingSphere-SQL - Actual SQL: database_team_default ::: select max(email_campaign_account_id)  from EMAIL_CAMPAIGN_TEAM_0 UNION ALL select max(email_campaign_account_id)  from EMAIL_CAMPAIGN_TEAM_1 UNION ALL select max(email_campaign_account_id)  from EMAIL_CAMPAIGN_TEAM_2
```

# IFNULL():
```
MySQL [hireez_sharding_db]> select IFNULL(max(email_campaign_account_id),0) from EMAIL_CAMPAIGN;
+------------------------------------------+
| IFNULL(max(email_campaign_account_id),0) |
+------------------------------------------+
|                                        0 |
|                                        0 |
|                                        0 |
+------------------------------------------+
3 rows in set (0.007 sec)
```

```
[INFO ] 2022-05-07 02:14:03.723 [ShardingSphere-Command-17] ShardingSphere-SQL - Logic SQL: select IFNULL(max(email_campaign_account_id),0) from EMAIL_CAMPAIGN
[INFO ] 2022-05-07 02:14:03.725 [ShardingSphere-Command-17] ShardingSphere-SQL - SQLStatement: MySQLSelectStatement(table=Optional.empty, limit=Optional.empty, lock=Optional.empty, window=Optional.empty)
[INFO ] 2022-05-07 02:14:03.725 [ShardingSphere-Command-17] ShardingSphere-SQL - Actual SQL: database_team_default ::: select IFNULL(max(email_campaign_account_id),0) from EMAIL_CAMPAIGN_TEAM_0 UNION ALL select IFNULL(max(email_campaign_account_id),0) from EMAIL_CAMPAIGN_TEAM_1 UNION ALL select IFNULL(max(email_campaign_account_id),0) from EMAIL_CAMPAIGN_TEAM_2
```

Maybe we need to aggregate the results again after finding the results in slices. Although this sounds very troublesome.

### LINKED ISSUE #18261: ShardingSphere-Proxy not support some mysql functions in result merge case
### Which version of ShardingSphere did you use?
shardingsphere master

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
ShardingSphere-Proxy

### Steps to reproduce the behavior,
```sql
select max(order_price) from t_order;
```
 execute log
```
Logic SQL: select max(order_price) from t_order
SQLStatement: MySQLSelectStatement(table=Optional.empty, limit=Optional.empty, lock=Optional.empty, window=Optional.empty)
Actual SQL: ds_0 ::: select max(order_price) from t_order_0 UNION ALL select max(order_price) from t_order_1
Actual SQL: ds_1 ::: select max(order_price) from t_order_0 UNION ALL select max(order_price) from t_order_1
```

execute result:
```
max(order_price)
32.0 
```

but when I add the function before max(order_price) as NULLIF(), rpad(),CONCAT() etc , shardingsphere-proxy return unmerged result;

like this,

#### NULLIF()
```
select NULLIF(max(order_price),0.0) from t_order;
```

log
```
Logic SQL: select NULLIF(max(order_price),0.0) from t_order
SQLStatement: MySQLSelectStatement(table=Optional.empty, limit=Optional.empty, lock=Optional.empty, window=Optional.empty)
Actual SQL: ds_0 ::: select NULLIF(max(order_price),0.0) from t_order_0 UNION ALL select NULLIF(max(order_price),0.0) from t_order_1
Actual SQL: ds_1 ::: select NULLIF(max(order_price),0.0) from t_order_0 UNION ALL select NULLIF(max(order_price),0.0) from t_order_1
```
execute result
```
NULLIF(max(order_price),0.0)
32.0
32.0
32.0
```
#### rpad()
```
select rpad(max(order_price),10,"$") from t_order;
```

execute log
```
Logic SQL: select rpad(max(order_price),10,"$") from t_order
Actual SQL: ds_0 ::: select rpad(max(order_price),10,"$") from t_order_0 UNION ALL select rpad(max(order_price),10,"$") from t_order_1
Actual SQL: ds_1 ::: select rpad(max(order_price),10,"$") from t_order_0 UNION ALL select rpad(max(order_price),10,"$") from t_order_1
```

execute result
```
rpad(max(order_price),10,"$")
32.00$$$$$
32.00$$$$$
32.00$$$$$
```

#### CONCAT()
```
select CONCAT(max(order_price),"$") from t_order;
```

execute log
```
Logic SQL: select CONCAT(max(order_price),"$") from t_order
SQLStatement: MySQLSelectStatement(table=Optional.empty, limit=Optional.empty, lock=Optional.empty, window=Optional.empty)
Actual SQL: ds_0 ::: select CONCAT(max(order_price),"$") from t_order_0 UNION ALL select CONCAT(max(order_price),"$") from t_order_1
Actual SQL: ds_1 ::: select CONCAT(max(order_price),"$") from t_order_0 UNION ALL select CONCAT(max(order_price),"$") from t_order_1
```
execute result
```
CONCAT(max(order_price),"$")
32.00$
32.00$
32.00$
```

### Actual behavior

the expected number of results is 1, like
```
select rpad(max(order_price),10,"$") from t_order;
---
rpad(max(order_price),10,"$")
32.00$$$$$
```
```
select rpad(max(order_price),10,"$") from t_order;
---
CONCAT(max(order_price),"$")
32.00$
```

### Reason analyze 

My analysis is as follows，

shardingsphere-proxy execute sql like this
```
select CONCAT(max(order_price),"$") from t_order;
```
the size of result  from execute engine is more than 1, then ss will call the merge engine to execute result merge. 
But **isNeedProcessGroupBy()** method line 85 in ShardingDQLResultMerger.java not return true, and result not merged.
```java
@Override
    public MergedResult merge() throws SQLException {
        if (1 == queryResults.size() && !isNeedAggregateRewrite(sqlStatementContext)) {
            return new IteratorStreamMergedResult(queryResults);
        }
        ……
        MergedResult mergedResult = build(queryResults, selectStatementContext, columnLabelIndexMap, database);
        return decorate(queryResults, selectStatementContext, mergedResult);
    }


private MergedResult build(final List<QueryResult> queryResults, final SelectStatementContext selectStatementContext,
                               final Map<String, Integer> columnLabelIndexMap, final ShardingSphereDatabase database) throws SQLException {
        ……
        if (isNeedProcessGroupBy(selectStatementContext)) {  // the condition should be true, but return false
            return getGroupByMergedResult(queryResults, selectStatementContext, columnLabelIndexMap, schema);
        } 
        ……
        return new IteratorStreamMergedResult(queryResults);
    }
```

### PR DESCRIPTION
**Pull request by iisheng**
Support #17402 .
Support #18261  but not NULLIF

Changes proposed in this pull request:
- ExpressionProjection add parameters field
- AggregationProjection alias field add setter method
- Add AggregationProjection constructor method
- Modify ProjectionsContext#getAggregationProjections
- Modify ProjectionEngine#createProjection
- Add sql case about MySQL ifnull
- Add more unit test for ProjectionsContext
- Add more unit test for ProjectionEngine
- Add MySQL dql integration test case

### CODE REVIEW

File: shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/impl/AggregationProjection.java [hunk 1]
```diff
@@ -45,7 +43,8 @@ public class AggregationProjection implements Projection {
     
     private final String innerExpression;
     
-    private final String alias;
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/impl/AggregationProjection.java:48]: Why remove final modifier?

**Review by strongduanmu — CHANGES_REQUESTED** [review]
@iisheng Please add unit test and integration test to cover this logic.

File: shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/impl/AggregationProjection.java (hunk 1 above)
Comment (iisheng) [reply shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/impl/AggregationProjection.java:48]: In general, "alias" was  initialized by constructor method. But when AggregationProjection belong to ExpressionProjection, I want to reset AggregationProjection's alias.

**Comment by iisheng** [comment]
> @iisheng Please add unit test and integration test to cover this logic.

OK

File: shardingsphere-test/shardingsphere-parser-test/src/main/resources/case/dml/select-special-function.xml [hunk 2]
```diff
@@ -211,4 +211,92 @@
             <simple-table name="t_order" start-index="29" stop-index="35" />
         </from>
     </select>
+    
+    <select sql-case-id="select_ifnull">
```
Comment (strongduanmu) [inline shardingsphere-test/shardingsphere-parser-test/src/main/resources/case/dml/select-special-function.xml:215]: This PR should not change the SQL parsing, so there is no need to add additional SQL Parser test cases.

File: shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/engine/ProjectionEngine.java [hunk 3]
```diff
@@ -125,8 +128,16 @@ private ColumnProjection createProjection(final ColumnProjectionSegment projecti
         return new ColumnProjection(owner, projectionSegment.getColumn().getIdentifier().getValue(), projectionSegment.getAlias().orElse(null));
     }
     
-    private ExpressionProjection createProjection(final ExpressionProjectionSegment projectionSegment) {
-        return new ExpressionProjection(projectionSegment.getText(), projectionSegment.getAlias().orElse(null));
+    private ExpressionProjection createProjection(final TableSegment table, final ExpressionProjectionSegment projectionSegment) {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/engine/ProjectionEngine.java:131]: We need add unit test and integration test to cover these new logics.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/engine/ProjectionEngine.java (hunk 3 above)
Comment (iisheng) [reply shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/engine/ProjectionEngine.java:131]: OK, I'll do that.

**Comment by strongduanmu** [comment]
@iisheng Please sync the latest master branch code.

**Review by strongduanmu — CHANGES_REQUESTED** [review]
Please fix ci error first.

File: shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/ProjectionsContext.java [hunk 4]
```diff
@@ -147,10 +148,28 @@ public List<AggregationProjection> getAggregationProjections() {
...
             }
+            if (each instanceof ExpressionProjection) {
+                addExpressionProjectionInsideAggregationProjection((ExpressionProjection) each, result);
+            }
         }
         return result;
     }
     
+    private void addExpressionProjectionInsideAggregationProjection(final ExpressionProjection expressionProjection, final List<AggregationProjection> result) {
+        Collection<Projection> parameters = expressionProjection.getParameters();
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/ProjectionsContext.java:159]: Why is the parameters parameter required in expressionProjection ?

File: shardingsphere-infra/shardingsphere-infra-binder/src/test/java/org/apache/shardingsphere/infra/binder/segment/select/projection/ProjectionsContextTest.java [hunk 5]
```diff
@@ -128,6 +144,13 @@ private AggregationProjection getAggregationProjection() {
         return new AggregationProjection(AggregationType.COUNT, "(column)", "c", mock(DatabaseType.class));
     }
     
+    private ExpressionProjection getContainsParametersExpressionProjection() {
+        AggregationProjection aggregationProjection = new AggregationProjection(AggregationType.MAX, "(status)", "IFNULL(MAX(status), 0)", mock(DatabaseType.class));
+        ExpressionProjection expressionProjection = new ExpressionProjection("IFNULL(MAX(status), 0)", null);
+        expressionProjection.getParameters().add(aggregationProjection);
+        return expressionProjection;
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-binder/src/test/java/org/apache/shardingsphere/infra/binder/segment/select/projection/ProjectionsContextTest.java:151]: Please return result here.

File: shardingsphere-infra/shardingsphere-infra-binder/src/test/java/org/apache/shardingsphere/infra/binder/segment/select/projection/engine/ProjectionEngineTest.java [hunk 6]
```diff
@@ -114,11 +120,33 @@ public void assertCreateProjectionWhenProjectionSegmentInstanceOfColumnProjectio
...
+        });
+    }
+    
+    private List<ExpressionSegment> getExpressionSegments() {
+        List<ExpressionSegment> list = new ArrayList<>();
+        FunctionSegment functionSegment = new FunctionSegment(7, 28, "IFNULL", "IFNULL(MAX(status), 0)");
+        AggregationProjectionSegment parameter = new AggregationProjectionSegment(14, 24, AggregationType.MAX, "(status)");
+        functionSegment.getParameters().add(parameter);
+        list.add(functionSegment);
+
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-binder/src/test/java/org/apache/shardingsphere/infra/binder/segment/select/projection/engine/ProjectionEngineTest.java:143]: Please remove useless blank line.

File: shardingsphere-infra/shardingsphere-infra-binder/src/test/java/org/apache/shardingsphere/infra/binder/segment/select/projection/engine/ProjectionEngineTest.java [hunk 7]
```diff
@@ -114,11 +120,33 @@ public void assertCreateProjectionWhenProjectionSegmentInstanceOfColumnProjectio
...
+        List<ExpressionSegment> list = new ArrayList<>();
+        FunctionSegment functionSegment = new FunctionSegment(7, 28, "IFNULL", "IFNULL(MAX(status), 0)");
+        AggregationProjectionSegment parameter = new AggregationProjectionSegment(14, 24, AggregationType.MAX, "(status)");
+        functionSegment.getParameters().add(parameter);
+        list.add(functionSegment);
+        
+        ExpressionSegment right = new LiteralExpressionSegment(25, 25, 1);
+        ExpressionSegment binaryOperationExpression = new BinaryOperationExpression(7, 25, parameter, right, "+", "IFNULL(status, 0)+1");
+        list.add(binaryOperationExpression);
+
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-binder/src/test/java/org/apache/shardingsphere/infra/binder/segment/select/projection/engine/ProjectionEngineTest.java:147]: Please remove useless blank line.

File: shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/asserts/segment/expression/ExpressionAssert.java [hunk 8]
```diff
@@ -378,8 +379,12 @@ public static void assertExpression(final SQLCaseAssertContext assertContext,
             ProjectionAssert.assertProjection(assertContext,
                     (ExpressionProjectionSegment) actual, expected.getExpressionProjection());
         } else if (actual instanceof AggregationProjectionSegment) {
+            ExpectedAggregationProjection expectedAggregationProjection = expected.getAggregationProjection();
```
Comment (strongduanmu) [inline shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/asserts/segment/expression/ExpressionAssert.java:382]: Please move this logic to assertProjection.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/asserts/segment/expression/ExpressionAssert.java (hunk 8 above)
Comment (iisheng) [reply shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/asserts/segment/expression/ExpressionAssert.java:382]: This logic is for get expected unit test object, should not move to assertProjection.

**Pull request closed without merging**